### PR TITLE
Allow non-leaf nodes to be routable

### PR DESCRIPTION
**Pull request by nathanboktae**
Currently in cherrytree, if you want `forums/1` to be routable, and `forums/1/posts/2` to be routable, you need to create a separate leaf node for `forums/1` that is *not* in the hiarchy of `forums/1/posts/2`, despite these being clear children in the URL. This poses two big problems:

1. Unintuititive and repetative. This caught me off guard that routes for parents were not generated, and I had to repeat myself by doing another route call with `.index` on and repeating options passed in (in `cherrytree-for-knockout`'s case, this is the knockout component that is specified twice and registered twice).
2. the DOM of the extra leaf node is destroyed and rerendered, klling intermediate state like scroll positions and widgets not aware of cherrytree, and causing flicker. If I have any chrome around `forums/1` it will be destroyed because there is no linkage declared between `forums/1` route and `forums/1/params/2` without going out of my way to crawl all the routes and comparing the real URLs.

It is a breaking change to register these by default (though it's still alpha, so it seems fine). I highly recommend it, but I could concede to have it be an explicit opt in. One option is to register these by default only if any options are specified for parent routes.

Also I am not quite sure on the naming of an option to opt out of a parent route having a route generated for it. `route`, `navigatable`, or `abstract` came to mind but didn't seem as good as `routable`.

### CODE REVIEW

**Comment by nathanboktae** [comment]
Thinking about it again, I like `abstract` (terminology I remember from `angular-ui-router`). If I say `routable: false`, it feels like it includes the children too. We all remember abstract from OOP, which this maps over just perfectly. Thoughts there?

**Comment by KidkArolis** [comment]
Hi, sorry I didn't have a chance yet to have a good read through your proposal.
However, my first thought is that the ".index" route is cherrytree's "solution" to this problem. By creating that leaf route you signify that this particular subroute is routable. If you make that route do no work except for perhaps setting a flag somewhere to indicate it's an index page - those 2 problems you're describing should go away.

In case of /forums/1 you would have something like

['app', 'forums', 'forums.index'] - where forums.index doesn't do much - it's just to make this url routable. forums is where all the rendering happens.

In case of /forums/1/posts/2

['app', 'forums', 'posts', 'posts.index'] - where 'posts.index' doesn't do much, all the rendering can be done by forums and posts this time.

As you navigate between the 2 - 'app' and 'forums' can stay as is, no teardown is needed. And tearing down 'forums.index' or 'posts.index' should be lightweight if they didn't render anything in the first place.

File: guide.md [hunk 1]
```diff
@@ -111,11 +90,8 @@ Route nesting is one of the core features of cherrytree. It's useful to nest rou
...
 router.map(function(route) {
   route('gmail', {path: '/'}, function () {
-    route('index')
     route('inbox', function() {
-      route('inbox.index')
       route('mail', {path: 'm/:mailId'}, function () {
-        route('mail.index')
         route('mail.raw')
       })
     })
```
Comment (nathanboktae) [inline guide.md]: See how much this got cleaned up now that we reflect the true nature of the URL

**Comment by nathanboktae** [comment]
> In case of /forums/1 you would have something like
>
> ['app', 'forums', 'forums.index'] - where forums.index doesn't do much - it's just to make this url routable. forums is where all the rendering happens.

But why do that when the natural thing is more simple?

> In case of /forums/1/posts/2
>
> ['app', 'forums', 'posts', 'posts.index'] - where 'posts.index' doesn't do much, all the rendering can be done by forums and posts this time.

 The problem here is that `/forums/1` maps to route `forums.index` when it's the leaf, but when it's a non-leaf it the `forum/1` maps to `posts`, so the appear different to anyone mapping a UI shell around that and it gets destroyed.

I think since you are using react with it's virtual DOM, you probably haven't experienced this, but this is deal breaker for me. I can't correlate `forum.index` and `posts`, beyond it's awkward and counter-intuitive.

**Comment by KidkArolis** [comment]
I'll have a closer look soon, these were just initial thoughts.

> The problem here is that /forums/1 maps to route forums.index when it's the leaf, but when it's a non-leaf it the forum/1 maps to posts, so the appear different to anyone mapping a UI shell around that and it gets destroyed.

I'm not following this. What do you mean `forum/1` maps to `posts`?

I'm actually using this with Backbone primarily and we do reconcile the differences efficiently by inspecting which routes are already active and not tearing those down using this middleware:

```js
 var _ = require('underscore')
var computeRelevantParams = require('../helpers/compute_relevant_params')

/**
 * mark up already active routes
 * so that we know if we should call model on them
 * as well as whether to we need to deactivate/activate them
 */
module.exports = function computeAlreadyActive (transition) {
  _.find(transition.routes, function (route, i) {
    var prevRoute = transition.prev.routes[i]
    if (prevRoute && routeInCorrectState(route, prevRoute)) {
      route.alreadyActive = true
      // continue looping
      return
    }
    // break
    return true
  })

  /**
   * Routes have handlers in mdashboard,
   * those handlers are stateful objects,
   * let's inspect if the handler is currently
   * in the right state.
   */
  function routeInCorrectState (route, prevRoute) {
    return prevRoute.name === route.name &&
    route.handler._currentParams !== null && // this would mean, the state is unknown
    _.isEqual(computeRelevantParams(transition, route), route.handler._currentParams)
  }
}
```

**Comment by nathanboktae** [comment]
> I'm not following this. What do you mean forum/1 maps to posts?

Never mind, I see from your middleware and previous example that `forums/1` would always map to post, and `forums.index` must really do nothing and be a stub for such middleware to work and not have flicker during a transition down the hierarchy or to adjacent children. BTW, I basically have the same code for comparing if the route is in the same state as before.

OK I will await your further inspection.

**Comment by msokk** [comment]
:+1: for this, as managing subviews gets complex otherwise

File: lib/router.js [hunk 2]
```diff
@@ -79,9 +79,10 @@ Cherrytree.prototype.map = function (routes) {
 
   function eachBranch (node, memo, fn) {
     node.routes.forEach(function (route) {
-      if (!route.routes || route.routes.length === 0) {
+      if (!route.options || route.options.abstract !== true) {
```
Comment (KidkArolis) [inline lib/router.js]: What would be the use case for the `abstract` route in this new version? Ideally it should work only one way (e.g. route to leaf routes only or route to any route only).

Comment (KidkArolis) [inline lib/router.js]: I'll try this version in my app an the example apps I have and see how it feels. Overall, looks like it could indeed be a simplification, especially judging by the fact that there are way less docs.

Comment (nathanboktae) [inline lib/router.js]: The only case for `abstract` is when listing routes to clearly indicate a level is not routable. Say `/account` defaults to `/account/profile`, then abstract could indicate that. However that could also be accomplished by registering each leaf with explict paths, like `path: "account/profile" `.

I like the idea of removing it - I'll do that.

Comment (nathanboktae) [inline lib/router.js]: Also users should always have a catch handler middleware so users don't end up stuck somewhere, and that or similar middleware can redirect for what would be the abstract non-leaf nodes.

**Comment by nathanboktae** [comment]
@KidkArolis Did you get a chance to go over the examples?

**Comment by KidkArolis** [comment]
Sorry for the lag. This is a tricky change, I don't want to rush into merging.

I would like to better understand your use case, because perhaps there's a different solution to this problem? In my case - I find the "only leaf routes are routable" restriction to be beneficial - this avoids that issue you brought up, users can never land on invalid, intermediate routes.

There are currently several uses for the non leaf routes:
- they could choose to not render anything and only be used for creating data structures, fetching data
- they could render some shell, e.g. outer navigation, a sidebar
- they could render an entire page - while this is typically done in leaf routes, you can also render entire page in the non leaf route in case your leafs only add extra content or modify something. In this case you'll need a blank index route. One example use case for this kind of route is if you want to have a subroute with a modal showing over the content of the page behind it...

Here's a more visual example of how I typically use cherrytree:

cherrytree example
(src: https://docs.google.com/drawings/d/1GGe4C88_ZS1U2Gyxy3HMhdaDTgWh8vgjcbo7icHj31Q/edit)

The only case that making non leafs routable in the above diagram is to avoid having the "editor.index" route. What if we invert this PR - you can opt into making non leaf routes routable with something like `{routable: true}` option? I'm suggesting this, because typically, most of the non leaf routes in my experience are not meant to be routable.

Another suggestion would be to have a cherrytree initialisation option `{routableNonLeafs: true}` or smth to then give the choice on the router level - to cater different use cases. Cherrytree is meant to be flexible and cater for different UI frameworks and different use cases.

What do you think?

OK. Also, I _will_ actually try to pull in this version into my large project - I do think it might simplify some of my stuff..

**Comment by KidkArolis** [comment]
Ok! I actually tried this in my app and it's awesome! I take everything back :)

Basically, I realised I used most of my '.index' routes to just redirect to some other route anyways. With this change - I can basically do the redirects in my non leaf routes using a simple redirect hook I implemented using a middleware (not including the entire middleware for brevity):

```js
...
function redirect (context) {
  // if handler has a redirect hook and it's the last route object in the current transition
  // call the redirect hook since this route doesn't want to be the 'final' route
  if (route.handler.redirect && transition.routes[transition.routes.length - 1].name === route.name) {
    return route.handler.redirect(params, context)
  }
  return context
}
...
```

This allowed me to remove 9 routes from the route map (and 9 files). Instead I moved all the code from those 9 files into the non leaf routes, into the redirect hook and can achieve the same outcome.

Basically, with this change then, it's up to you what to do when someone lands on the non leaf route - you can keep the user there, or you can redirect them to a deeper route (many ways to do that using middleware, depending on how the app is structured).

I've found 1 small edge case that would be good address though before we could merge this PR. I'll post it in a separate comment for clarity.

**Comment by KidkArolis** [comment]
Ok. So as I mentioned above - I quite like this idea of allowing non leafs to be routable. However, I still want to keep a possibility of having 'index' like routes. For example:

```js
route('app', {path: '/app'}, function () {
  route('dashboard', {path: ''})
  route('realtime')
  route('settings')
})
```

In this example, I want to be able to visit:

```
/app
/app/realtime
/app/settings
```

However, 2 routes now have identical path:

```
app - /app
dashboard - /app
```

So, I was thinking, in this case, we _could_ automatically treat `app` route as abstract, since it basically clearly we want to load route `dashboard` whenever someone goes to `/app`. Do you agree with this proposal?

However, there is only one small issue with this. You can't `router.generate` or `router.transitionTo` those 'abstract' urls, which can be confusing. Normally I can transition to any route (leaf or non leaf), but depending on how we implement it, I won't be able to transition to this abstract/shadowed route. I see two options there when someone transitions/generates a route that is shadowed:

* we could throw a clear error - "Route 'app' is abstract and can not be transitioned to" or "Route 'app' is shadowed by 'dashboard', please transition to 'dashboard' instead"
* we could allow transitioning, but automatically transition to the "matching" route, e.g. `router.transitionTo('app')` is equivalent to `router.transitionTo('dashboard')`

Currently I'm actually leaning towards option 1.. That way you don't have multiple names matching the same route, which could be important if you use the route name for something like analytics.

This all is kind of what you suggested earlier on with 'abstract' option, but instead:

* the behaviour is automatic (in case of `path: ''`)
* if you want a routable route to be "abstract" - you redirect within it to another route...

Finally, if someone really wants to mark many routes as "abstract", you can set that option and create a simple middleware that checks if that option is set and if this route was transitioned to - you can redirect to a 404 or wherever. Basically, this can be solved in userland via a middleware!

**Comment by nathanboktae** [comment]
> Ok! I actually tried this in my app and it's awesome! I take everything back :)

hehe Great! People that are analytic and articulate but humble to truly consider other opinions are rare!

Yup I realized that anyone can change the URL to anything, so you'll need to handle bad URLs (redirect somewhere else, for example) anyways whether or not Cherrytree considers them routable or not. In my app I'm working to make dialogs routable, so the general close handler wants to just go back, e.g. up a level. If that upper level doesn't make sense, it should be someone else's job to redirect.

So Option 1 does look better, and I think know Cherrytree has that issue where users can declare routes that match the same path, so it needs to be dealt with anyways.

However, what about declaring those routes like this?

```
route('dashboard', {path: '/app'}, function () {
  route('realtime')
  route('settings')
})
```

Is it because there may be additional stuff you want to do at `/app` that you don't want to do at `/app/settings` ? So middleware knows if it is a leaf node or not, so that can still support those scenarios

```
router.use(function(transition) {
  routeWebComponents = transition.routes.map(function(r, idx) {
    var component = // load React class/Polymer web component/knockout component/your fravorite framework thingie
    component.abstract = idx === transition.routes.length - 1
    return component
  })
})
```

But again, because cherrytree needs to handle conflicting paths either way, you do need to make a decision between 1 and 2.

**Comment by KidkArolis** [comment]
By the way, to answer your question and to continue the discussion - you asked why I don't do this:

```js
route('dashboard', {path: '/app'}, function () {
  route('realtime')
  route('settings')
})
```

Well, I do this

```js
route('app', {path: '/app'}, function () {
  route('dashboard', {path: ''})
  route('realtime')
  route('settings')
})
```

Because my `app` route is indeed abstract and it only prepares some data models and similar things that are then used in all three subroutes. So basically, I want them all to share that parent route. And then, each one, dashboard, realtime and settings render a full page UI (except nav, which is what app also renders). Furthermore, when I navigate `dashboard` -> `realtime` - I want to keep app as is (I mark it as active and don't touch it), but I want to then completely cleanup and unload `dashboard` route+views and instead load in the `realtime` route+views.

I suppose, this depends on how the middleware are implemented. In my case, middleware, for each route loads a 'route handler object' with methods such as `model`, `activate` and `deactivate`. So, basically, I want to call `deactivate` on `dashboard` when navigating away. Perhaps, you can achieve the same with your suggested route map by detecting if `dashboard` is last in the active route list or not and do stuff depending on that.

**Comment by KidkArolis** [comment]
By "do stuff" I mean - if `dashbaord` is the last route - render dashboard, if not - deactivate it and leave space for the subroute.

**Comment by KidkArolis** [comment]
Closing in favour of #100. Thanks for the PR and the discussion!
